contracts/ATokenYieldSource.sol

```diff
@@ -0,0 +1,408 @@
+    }
+
+    emit ApprovedERC20(msg.sender, _spender, _amount, _token);
+  }
+
+  /**
+   * @notice Transfer ERC20 tokens other than the aTokens held by this contract to the recipient address.
+   * @dev This function is only callable by the owner or asset manager.
+   * @param _token The ERC20 token to transfer
+   * @param _to The recipient of the tokens
+   * @param _amount The amount of tokens to transfer
+   */
+  function transferERC20(
+    IERC20 _token,
+    address _to,
+    uint256 _amount
+  ) external onlyManagerOrOwner {
+    require(address(_token) != address(aToken), "ATokenYS/forbid-aToken-transfer");
+    _token.safeTransfer(_to, _amount);
+    emit TransferredERC20(msg.sender, _to, _amount, _token);
+  }
+
+  /* ============ Internal Functions ============ */
+
+  /**
+   * @notice Calculates the number of shares that should be minted or burnt when a user deposit or withdraw.
+   * @param _tokens Amount of asset tokens
+   * @return Number of shares.
+   */
+  function _tokenToShares(uint256 _tokens) internal view returns (uint256) {
+    uint256 _supply = totalSupply();
+
+    // shares = tokens * (totalShares / yieldSourceATokenTotalSupply)
+    return _supply == 0 ? _tokens : _tokens.wadMul(_supply.wadDiv(aToken.balanceOf(address(this))));
+  }
+
+  /**
+   * @notice Calculates the number of asset tokens a user has in the yield source.
+   * @param _shares Amount of shares
+   * @return Number of asset tokens.
+   */
+  function _sharesToToken(uint256 _shares) internal view returns (uint256) {
+    uint256 _supply = totalSupply();
+
+    // tokens = (shares * yieldSourceATokenTotalSupply) / totalShares
+    return _supply == 0 ? _shares : _shares.mul(aToken.balanceOf(address(this))).div(_supply);
+  }
+
+  /**
+   * @notice Deposits asset tokens into the yield source.
+   * @param _assetToken ERC20 asset token address
+   * @param _amount Amount of asset tokens to be deposited
+   */
+  function _depositAssetTokens(IERC20 _assetToken, uint256 _amount) internal {
+    _assetToken.safeTransferFrom(msg.sender, address(this), _amount);
+  }
+
+  /**
+   * @notice Supply asset tokens to Aave.
+   * @param _mintAmount Amount of asset tokens to be supplied
+   */
+  function _supplyToAave(uint256 _mintAmount) internal {
+    address _underlyingAssetAddress = _tokenAddress();
+
+    _depositAssetTokens(IERC20(_underlyingAssetAddress), _mintAmount);
+    _pool().supply(_underlyingAssetAddress, _mintAmount, address(this), REFERRAL_CODE);
+  }
+
+  /**
+   * @notice Returns the underlying asset token address.
+   * @return Underlying asset token address.
+   */
+  function _tokenAddress() internal view returns (address) {
+    return aToken.UNDERLYING_ASSET_ADDRESS();
+  }
+
+  /**
+   * @notice Retrieves Aave PoolAddressesProvider address.
+   * @return A reference to PoolAddressesProvider interface.
+   */
+  function _poolProvider() internal view returns (IPoolAddressesProvider) {
```

Comment:
Yes, I was wondering this too while coding but since it's an internal function and there is the natspec doc to explain what it's used for, I think it's better to have a shorter name.